### PR TITLE
QF2050-select_provider_email_apim

### PR DESCRIPTION
**Pull request by karen-hanna**

### CODE REVIEW

**Review by ChrisJWoodcock — CHANGES_REQUESTED** [review]

File: src/ProviderRequestApprenticeTraining/SFA.DAS.ProviderRequestApprenticeTraining.Api/AppStart/AddServiceRegistrationExtension.cs [hunk 1]
```diff
@@ -20,6 +20,7 @@ public static void AddServiceRegistration(this IServiceCollection services)
 
             services.AddTransient<IRoatpCourseManagementApiClient<RoatpV2ApiConfiguration>, RoatpCourseManagementApiClient>();
             services.AddTransient<IRequestApprenticeTrainingApiClient<RequestApprenticeTrainingApiConfiguration>, RequestApprenticeTrainingApiClient>();
+            services.AddTransient<IProviderCoursesApiClient<ProviderCoursesApiConfiguration>, ProviderCoursesApiClient>();
```
[the remark this answers is not in the document]
Comment (karen-hanna) [reply src/ProviderRequestApprenticeTraining/SFA.DAS.ProviderRequestApprenticeTraining.Api/AppStart/AddServiceRegistrationExtension.cs:23]: removed

File: src/ProviderRequestApprenticeTraining/SFA.DAS.ProviderRequestApprenticeTraining.Api/AppStart/AddConfigurationOptionsExtension.cs [hunk 2]
```diff
@@ -15,9 +16,11 @@ public static void AddConfigurationOptions(this IServiceCollection services, ICo
 
             services.Configure<RoatpV2ApiConfiguration>(configuration.GetSection(nameof(RoatpV2ApiConfiguration)));
             services.Configure<RequestApprenticeTrainingApiConfiguration>(configuration.GetSection(nameof(RequestApprenticeTrainingApiConfiguration)));
+            services.Configure<ProviderCoursesApiConfiguration>(configuration.GetSection(nameof(RoatpV2ApiConfiguration)));
```
[the remark this answers is not in the document]
Comment (karen-hanna) [reply src/ProviderRequestApprenticeTraining/SFA.DAS.ProviderRequestApprenticeTraining.Api/AppStart/AddConfigurationOptionsExtension.cs:19]: removed

File: src/ProviderRequestApprenticeTraining/SFA.DAS.ProviderRequestApprenticeTraining.Api/AppStart/AddConfigurationOptionsExtension.cs [hunk 3]
```diff
@@ -15,9 +16,11 @@ public static void AddConfigurationOptions(this IServiceCollection services, ICo
...
 
             services.AddSingleton(cfg => cfg.GetService<IOptions<RoatpV2ApiConfiguration>>().Value);
             services.AddSingleton(cfg => cfg.GetService<IOptions<RequestApprenticeTrainingApiConfiguration>>().Value);
+            services.AddSingleton(cfg => cfg.GetService<IOptions<ProviderCoursesApiConfiguration>>().Value);
```
[the remark this answers is not in the document]
Comment (karen-hanna) [reply src/ProviderRequestApprenticeTraining/SFA.DAS.ProviderRequestApprenticeTraining.Api/AppStart/AddConfigurationOptionsExtension.cs:23]: removed

File: src/ProviderRequestApprenticeTraining/SFA.DAS.ProviderRequestApprenticeTraining.Api/Controllers/EmployerRequestsController.cs [hunk 4]
```diff
@@ -116,6 +117,25 @@ public async Task<IActionResult> UpdateProviderResponseStatus(CreateProviderResp
             }
         }
 
+        [HttpGet("provider/{ukprn}/emails")]
```
[the remark this answers is not in the document]
Comment (karen-hanna) [reply src/ProviderRequestApprenticeTraining/SFA.DAS.ProviderRequestApprenticeTraining.Api/Controllers/EmployerRequestsController.cs:120]: updated

**Review by ChrisJWoodcock — CHANGES_REQUESTED** [review]

File: src/ProviderRequestApprenticeTraining/SFA.DAS.ProviderRequestApprenticeTraining/Application/Queries/GetProviderEmailAddresses/GetProviderEmailAddressesQueryHandler.cs [hunk 5]
```diff
@@ -16,39 +17,35 @@ namespace SFA.DAS.ProviderRequestApprenticeTraining.Application.Queries.GetProvi
...
-            var providerCourses = await _providerCoursesApiClient.
+            var providerCourses = await _roatpCourseManagementApiClient.
                 Get<List<ProviderCourse>>(new GetProviderCoursesRequest(
                     request.Ukprn));
 
-            var providerSummary = await _roatpCourseManagementApiClient.
-                Get<GetProviderSummaryResponse>(new GetRoatpProviderRequest(
-                    Convert.ToInt32(request.Ukprn)));
+            var emailAddresses = (providerCourses?.Where(pc => !string.IsNullOrWhiteSpace(pc.ContactUsEmail))
+                .Select(pc => pc.ContactUsEmail.RemoveWhitespace()) ?? Enumerable.Empty<string>())
```
[the remark this answers is not in the document]
Comment (karen-hanna) [reply src/ProviderRequestApprenticeTraining/SFA.DAS.ProviderRequestApprenticeTraining/Application/Queries/GetProviderEmailAddresses/GetProviderEmailAddressesQueryHandler.cs:39]: So ( i think) line 38 allows for providercourses to be null and selects all non-empty strings.  Them line 39 selects the results of that and returns an empty enumerable<stsring> if there are no results (i.e. if the providercourses was null or empty).

File: src/ProviderRequestApprenticeTraining/SFA.DAS.ProviderRequestApprenticeTraining/Application/Queries/GetProviderEmailAddresses/GetProviderEmailAddressesQueryHandler.cs [hunk 6]
```diff
@@ -16,39 +17,35 @@ namespace SFA.DAS.ProviderRequestApprenticeTraining.Application.Queries.GetProvi
...
+                .Append(request.UserEmailAddress.RemoveWhitespace() ?? string.Empty)
+                .Where(phone => !string.IsNullOrWhiteSpace(phone))
```
[the remark this answers is not in the document]
Comment (karen-hanna) [reply src/ProviderRequestApprenticeTraining/SFA.DAS.ProviderRequestApprenticeTraining/Application/Queries/GetProviderEmailAddresses/GetProviderEmailAddressesQueryHandler.cs]: updated to email

File: src/ProviderRequestApprenticeTraining/SFA.DAS.ProviderRequestApprenticeTraining/Application/Queries/GetProviderEmailAddresses/GetProviderEmailAddressesQueryHandler.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (ChrisJWoodcock) [reply src/ProviderRequestApprenticeTraining/SFA.DAS.ProviderRequestApprenticeTraining/Application/Queries/GetProviderEmailAddresses/GetProviderEmailAddressesQueryHandler.cs:39]: I read that as if providerCourses is empty it return null and does a .Select on the null.

var emailAddresses = (providerCourses?
	.Select(pc => pc.ContactUsEmail) ?? Enumerable.Empty<string>())
	.Append(request.UserEmailAddress)
	.Select(email => email.RemoveWhiteSpace())
	.Where(email => !string.IsNullOrEmpty(email))
	.Distinct()
	.Order()
	.ToList();

**Review by ChrisJWoodcock — CHANGES_REQUESTED** [review]

File: src/ProviderRequestApprenticeTraining/SFA.DAS.ProviderRequestApprenticeTraining/Application/Queries/GetProviderEmailAddresses/GetProviderEmailAddressesQueryHandler.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (ChrisJWoodcock) [reply src/ProviderRequestApprenticeTraining/SFA.DAS.ProviderRequestApprenticeTraining/Application/Queries/GetProviderEmailAddresses/GetProviderEmailAddressesQueryHandler.cs:39]: See comment below

[the remark this answers is not in the document]
Comment (karen-hanna) [reply src/ProviderRequestApprenticeTraining/SFA.DAS.ProviderRequestApprenticeTraining/Application/Queries/GetProviderEmailAddresses/GetProviderEmailAddressesQueryHandler.cs:39]: i will defer to your authority on this matter

**Review by ChrisJWoodcock — DISMISSED** [review]

**Review by ChrisJWoodcock — APPROVED** [review]

**Review by ChrisJWoodcock — APPROVED** [review]

**Pull request merged**